R/js.R

```diff
@@ -0,0 +1,141 @@
+#' Construct JavaScript method to capture Shiny inputs on change.
+#'
+#' This captures the state of a `sortable` list.  Typically you would use this
+#' with the `onSort` option of `sortable_js`. See [sortable_options()].
+#'
+#' @param input_id Shiny input name to set
+#'
+#' @seealso [sortable_js] and [rank_list].
+#'
+#' @rdname sortable_js_capture_input
+#' @export
+#' @examples
+#' # For an example, see the Shiny app at
+#' system.file("shiny-examples/drag_vars_to_plot/drag_vars_to_plot_app.R", package = "sortable")
+sortable_js_capture_input <- function(input_id) {
+  # can call jquery as shiny will always have jquery
+  inner_text <- "$.map(this.el.children, function(child){return child.innerText})"
+  js_text <- "function(evt) {
+  if (typeof Shiny !== \"undefined\") {
+    Shiny.setInputValue(\"%s:sortablejs.rank_list\", %s)
+  }
+}"
+
+  js <- sprintf(js_text, input_id, inner_text)
+
+  htmlwidgets::JS(js)
+}
+
+
+#' @rdname sortable_js_capture_input
+#' @param input_ids Set of Shiny input ids to set corresponding to the provided `selectors`
+#' @param selectors Set of SortableJS selector values to help retrieve all to set as an object
+#' @export
+sortable_js_capture_bucket_input <- function(input_id, input_ids, selectors) {
+  assert_that(length(input_ids) > 0)
+  assert_that(length(input_ids) == length(selectors))
+
+  # can use jquery as shiny will have jquery
+  js_text <- "function(evt) {
+  if (typeof Shiny == \"undefined\") {
+    return;
+  }
+
+  var ret = {}, i;
+  var selectors = %s;
+  var input_ids = %s;
+
+  $.map(selectors, function(selector, i) {
+    var input_id = input_ids[i];
+    var item = $('#' + selector).get(0);
+    if (item && item.children) {
+      ret[input_id] = $.map(item.children, function(child) {
+        return child.innerText;
+      });
+    } else {
+      ret[input_id] = undefined;
+    }
+  });
+  Shiny.setInputValue(\"%s:sortablejs.bucket_list\", ret)
+}"
+
+  js <- sprintf(
+    js_text,
+    base::toString(jsonlite::toJSON(selectors)),
```

Comment:
`selectors` should be a list to make sure that single-element objects are converted to JS arrays.

Reply:
Also, use `as.character` instead of `toString`.